### DISCUSSION TITLE
Serialize query from/to JSON

### DISCUSSION THREAD

**Discussion by HamedFathi** [alirezanet/Gridify#106] [Q&A] [answered]
Hi,

First, thank you so much for such a great piece of work.

Second, before getting familiar with your library I worked on my own [here](https://github.com/HamedFathi/Beyond.QueryBuilder). and just like [jQuery QueryBuilder](https://querybuilder.js.org/demo.html) if you click on one of the `Get Rule` buttons I used an intermediate representation as a C# model (also JSON) to be able to pass entire configs, rules, and filters via network and convert it as C# expression again.

![image](https://github.com/alirezanet/Gridify/assets/8418700/007d91c4-67da-4b94-bf59-54db58b1c4a0)

![image](https://github.com/alirezanet/Gridify/assets/8418700/f83f9cc7-615b-4a57-82d9-fb7f2da5e651)

![image](https://github.com/alirezanet/Gridify/assets/8418700/7fcb00b5-20df-4027-a4e7-d105b894f04b)

Is it possible to have a similar intermediate representation with all nested levels, index mapping, and available information as a model (JSON) for expression? (converting query from/to expression and passing it through a network or saving it in DB to restore in the future)

I think this also helps us to have our own query builder UI on the FE side to work with this library in BE.

For example consider `JsonLogic` in [react-awesome-query-builder](https://github.com/ukrbublik/react-awesome-query-builder)

![image](https://github.com/alirezanet/Gridify/assets/8418700/7e9bcf5e-523a-4090-b300-312620d19d23)

What do you think? Is it possible?

**Comment by alirezanet** [comment, marked as answer]
Hi @HamedFathi,
Currently we don't have a built-in parser for Json deserialization but I think even if Gridify doesn't support it directly you can create a helper class to parse and convert json to a gridify query expression. but if you have a better suggestion happy to hear that.

**Reply by HamedFathi** [reply]
Hi Alireza,

I knew that, I want to replace my library with yours but I asked to make sure writting a wrapper on top of Gridify is a good option.

Thanks.

**Reply by alirezanet** [reply]
You're welcome, Yes, I think it works but if the performance is too important to you, I believe I can add a feature to support different parsers... let me know if you faced any problems...

**Reply by HamedFathi** [reply]
Lovely, yes, adding a feature for a custom parser is brilliant, even for visitors in your AST if possible.

Another question, Is it possible to use the current parser for the string we pass? Are they available as public APIs? I know it does not make sense but just a question. I might want to split your DSL language to a JSON for any reason. Can I parser them with the built-in parser too?

I also read more discussions about which developers want to use this great library for FE. We can finally help them via this thread, and lead us to a good solution.

Anyway, thanks.

**Reply by alirezanet** [reply]
>Another question, Is it possible to use the current parser for the string we pass? Are they available as public APIs? I know it does not make sense but just a question. I might want to split your DSL language to a JSON for any reason. Can I parser them with the built-in parser too?

The [parser](https://github.com/alirezanet/Gridify/blob/master/src/Gridify/Syntax/Parser.cs) and lexer are internal classes but you can access the generated linq expression using the query builder ... I didn't understand what you mean exactly since I don't know your use case well, so it would be nice if you provide an example of what you need or what you wanna do so I can help you or at least answer your questions better. 😊

**Reply by alirezanet** [reply]
> Lovely, yes, adding a feature for a custom parser is brilliant, even for visitors in your AST if possible.

I agree, would be a cool feature. I'll try to take a look later

**Reply by HamedFathi** [reply]
Honestly, I was thinking loudly! Please inform me if you have a new release about custom parsers and visitors.

Thanks.
